### ISSUE TITLE
How to estimate operational intensity

### ISSUE THREAD

**Issue by brnorris03** [betterscientificsoftware/betterscientificsoftware.github.io#14]
The operational intensity introduced in the [Roofline model](https://escholarship.org/uc/item/5tz795vq.pdf) -- operations per byte of DRAM traffic -- is a simple model that can be used to determine what architectures are the best match for a given computational kernel, or conversely, in what ways to optimize a kernel so it performs better on a given architecture. Operational intensity is not typically provided directly by performance tools but can be estimated from other readily available measurements.

**Comment by ibaned** [comment]
I've been told that this project provide hardware-independent estimates by parsing and instrumenting the user's code:

https://github.com/lanl/Byfl

**Comment by brnorris03** [comment]
Yes, it's one of the approaches (software performance counters through binary instrumentation). Specific estimates would necessarily be hardware-specific but the tool is not architecture specific as far as it's based on LLVM intermediate code instrumentation. There are two others -- hardware counter-based (architecture-specific), or high-level source code analysis (hardware-parametric). These can be combined in various hybrid approaches.

**Comment by ibaned** [comment]
I'd be interested in examples of high-level source code analysis tools.
